slowness networking stack - MXRT1050-EVK

I'm trying the "samples/net/sockets/dumb_http_server", i just developed a little webserver based on this example.
The request always takes about 350ms, you can see it from the browser console (F12 -> network).
I think the problem is not related to the board or network because my bare-metal version is much faster (the HTTP respose is like 1ms).

I think the problem is in the "accept" call, but I don't know if is the context switch delay from the "idle" thread or other.

Can anyone reproduce the problem? I'm using the MIMXRT1050-EVK.

Ok, so I'm bisecting this to see if it can help to understand the delay cause. As we know, 338dc8a9520ea7b3c60503e35daba085073f4a1d works w/o delay. The last commit of that "streak" which works is 3215dd862f93c68a0a96ad56e42ae4edb80b5253.

In the next commit, f484bbaa264f3a9640ca52477e3038f912ea5642, I myself broke the build of dumb_http_server (`fdtable.c:80:21: error: ‘FD_OBJ_RESERVED’ undeclared (first use in this fu)
   fdtable[fd].obj = FD_OBJ_RESERVED;`).

This error happens until (and including) 9aafe388280f5920b9400cc4c3b62fc548a2eeb6. In the next commit, 65a33bd0c07a29e71630f910d6304f2a0468f775, the build is fixed, but runtime of the example is broken (`Error in accept: 23 - continuing`). Last commit with that error is aa5bbd5dfe9930d26555d0db95df0291386a4129.

Then it recovered in 2e21a95fd470e9f244e7532dae9b69f181f9ceca and serves w/o delay again! It works until and including df07e093ce5e7696705f1408e315432fb9275a58. Next commit, 94e89a11b7151c9f8dd8c22007e64bd30ee8b809, is beginning of @tbursztyka's last refactor (it was split after a number of PRs after. e.g. https://github.com/zephyrproject-rtos/zephyr/pull/10547 is one of them). The 94e89a11b7151c9f8dd8c22007e64bd30ee8b809 still works for dumb_http_server. Then with next commit, dcc6eddfefe51eacd2ecd60d1c2596b1ceea338a, packets don't go thru the stack. dump_http_server again works w/o delay at 10c4841835249bfc61d47f18f1908e77fca77377, the end of this streak of @tbursztyka's changes!


(to be continued)

--- a/subsys/net/ip/tcp.c
+++ b/subsys/net/ip/tcp.c
@@ -972,6 +972,11 @@ int net_tcp_send_pkt(struct net_pkt *pkt)
 	return net_send_data(pkt);
 }
 
+static void flush_queue(struct net_context *context)
+{
+	(void)net_tcp_send_data(context, NULL, NULL);
+}
+
 static void restart_timer(struct net_tcp *tcp)
 {
 	if (!sys_slist_is_empty(&tcp->sent_list)) {
@@ -1134,6 +1139,11 @@ bool net_tcp_ack_received(struct net_context *ctx, u32_t ack)
 	 */
 	if (valid_ack) {
 		restart_timer(ctx->tcp);
+
+		/* Flush anything pending. This is important as if there
+		 * is FIN waiting in the queue, it gets sent asap.
+		 */
+		flush_queue(ctx);
 	}
 
 	return true;
@@ -1377,6 +1387,7 @@ int net_tcp_recv(struct net_context *context, net_context_recv_cb_t cb,
 static void queue_fin(struct net_context *ctx)
 {
 	struct net_pkt *pkt = NULL;
+	bool flush = false;
 	int ret;
 
 	ret = net_tcp_prepare_segment(ctx->tcp, NET_TCP_FIN, NULL, 0,
@@ -1385,7 +1396,15 @@ static void queue_fin(struct net_context *ctx)
 		return;
 	}
 
+	if (sys_slist_is_empty(&ctx->tcp->sent_list)) {
+		flush = true;
+	}
+
 	net_tcp_queue_pkt(ctx, pkt);
+
+	if (flush) {
+		flush_queue(ctx);
+	}
 }
 
 int net_tcp_put(struct net_context *context)

